datafusion/optimizer/src/optimizer.rs

```diff
@@ -357,6 +357,7 @@ impl Optimizer {
     {
         let start_time = Instant::now();
         let options = config.options();
+        plan.validate_parameter_types()?;
```

Comment:
Checking parameter types doesn't seem to logically belong to the optimizer which is supposed to preserve the semantics of each plan but potentially make them faster.

https://docs.rs/datafusion/latest/datafusion/optimizer/trait.OptimizerRule.html

> OptimizerRules transforms one [LogicalPlan](https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.LogicalPlan.html) into another which computes the same results, but in a potentially more efficient way. If there are no suitable transformations for the input plan, the optimizer should simply return it unmodified.

I think the `Analyzer` might be a more appropriate place to put this check:

https://docs.rs/datafusion/latest/datafusion/optimizer/struct.Analyzer.html#method.execute_and_check

Reply:
Although thinking about it some more, if we are going to defer the error anyways, I think the cleanest solution would be to move the error to when it is needed

When I removed the specific check in the Optimizer and tried to run the queries, it failed with this error:

```
This feature is not implemented: Physical plan does not support logical expression Placeholder(Placeholder { id: "$1", data_type: None })
```

Maybe rather than a special check we could simply make this error better

Reply:
(we can do this in a follow on PR)

Reply:
I thought your suggestion was a lot cleaner and have pushed e5325c3f4 to address it. I've updated the dataframe tests, added a few to the core/tests/sql/select.rs module, and added a `PREPARE` test to the sqllogic test showing how things fail there as well.

Let me know if you can think of anywhere else I should add tests.